File: backend_golang/presentation/handler/get_todo_handler.go
package handler

import (
	"backend_golang/application_service/usecase"
	"backend_golang/domain/repository"
	"errors"
	"net/http"

	"github.com/google/uuid"
	"github.com/labstack/echo/v4"
)

type GetTodoHandler struct {
	gu usecase.GetTodoUsecase
}

func NewGetTodoHandler(gu usecase.GetTodoUsecase) HTTPHandler {
	return &GetTodoHandler{
		gu: gu,
	}
}

type GetTodoRequest struct {
	ID uuid.UUID `param:"id" validate:"required"`
}

type GetTodoResponse struct {
	ID          uuid.UUID `json:"id"`
	Title       string    `json:"title"`
	Description *string   `json:"description"`
	Completed   bool      `json:"completed"`
}

func (h *GetTodoHandler) Handle(c echo.Context) error {
	var req GetTodoRequest
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}
	if err := c.Validate(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}
	todo, err := h.gu.Handle(c.Request().Context(), req.ID)
	if err != nil {
		if errors.Is(err, repository.ErrRepositoryNotFound) {
			return echo.NewHTTPError(http.StatusNotFound, "Todo not found")
		}
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	t := todo.Serialize()
	return c.JSON(http.StatusOK, GetTodoResponse{
		ID:          t.ID,
		Title:       t.Title,
		Description: t.Description,
		Completed:   t.Completed,
	})
}

File: backend_bun/src/presentation/handler/getTodoHandler.ts
import { zValidator } from "@hono/zod-validator";
import type { Hono } from "hono";
import z from "zod";
import type { GetTodoUsecase } from "@/application_service/usecase/getTodoUsecase";
import { RepositoryNotFoundError } from "@/domain/repository/errors";
import {
  BadRequestError,
  InternalServerError,
  NotFoundError,
} from "@/presentation/middleware/errorHandler";

const getTodoRequestParamSchema = z.object({
  id: z.uuid({ version: "v7" }),
});

export class GetTodoHandler {
  constructor(private getTodoUsecase: GetTodoUsecase) {}

  register = (app: Hono) => {
    app.get(
      "/todos/:id",
      zValidator("param", getTodoRequestParamSchema, (result) => {
        if (!result.success) {
          throw new BadRequestError("Invalid request data", result.error);
        }
      }),
      async (c) => {
        try {
          const { id } = c.req.valid("param");
          const result = await this.getTodoUsecase.execute({
            id,
          });
          return c.json({
            id: result.todo.id,
            title: result.todo.title,
            description: result.todo.description,
            completed: result.todo.isCompleted,
          });
        } catch (error) {
          switch (true) {
            case error instanceof z.ZodError: {
              const message = error.issues
                .map((issue) => `${issue.path}: ${issue.message}`)
                .join(", ");
              throw new BadRequestError(
                `Invalid request data: ${message}`,
                error,
              );
            }
            case error instanceof RepositoryNotFoundError:
              throw new NotFoundError("Todo not found");
          }
          throw new InternalServerError("Failed to get todo", error);
        }
      },
    );
  };
}
